### PR TITLE
Outside Controller Documentation Use

### PR DESCRIPTION
**Pull request by CodedBeardedSignedTaylor**
Fulfilling #1036.

### CODE REVIEW

**Comment by CodedBeardedSignedTaylor** [comment]
Do we want to add a step of saying `serializer.to_json` for `SerializableResource.serialize`?

File: docs/howto/outside_controller_use.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+fairly simple:
+
+```ruby
+# Create our resource
+post = Post.create(title: "Sample post", body: "I love Active Model Serializers!")
+
+# Optional options parameters
+options = {}
+
+# Create the serializer!
```
Comment (beauby) [inline docs/howto/outside_controller_use.md]: Maybe it's better to avoid punctuation here.

File: docs/howto/outside_controller_use.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+```
+
+This will return an instance of the SerialzableResource class that will allow you to serialze your model. 
+
+### Retrieving a Resource's Active Model Serializer
+
+If you want to retrieve a serializer for a specific resource, you can do the following:
+
+```ruby
+# Creating that resource again.
```
Comment (beauby) [inline docs/howto/outside_controller_use.md]: Maybe it's better to avoid punctuation here as well, for consistency.

File: docs/howto/outside_controller_use.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+post = Post.create(title: "Another Example", body: "So much fun.")
+
+# Optional options parameters
+options = {}
+
+# Create the serializer!
```
Comment (beauby) [inline docs/howto/outside_controller_use.md]: And same here.

**Comment by CodedBeardedSignedTaylor** [comment]
Fixed.

File: docs/howto/outside_controller_use.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+post = Post.create(title: "Sample post", body: "I love Active Model Serializers!")
+
+# Optional options parameters
+options = {}
+
+# Create the serializer
+serializer = ActiveModel::SerializableResource.serialize(post, options)
+```
+
+This will return an instance of the SerialzableResource class that will allow you to serialze your model.
```
Comment (beauby) [inline docs/howto/outside_controller_use.md]: As a reader, I'm left with an `ActiveModel::SerializableResource` although I was expecting information to make an actual json string.

**Comment by CodedBeardedSignedTaylor** [comment]
Definitely. The initial issue referenced that step as well.

File: docs/howto/outside_controller_use.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+options = {}
+
+# Create the serializer
+serializer = ActiveModel::SerializableResource.serialize(post, options)
+
+# Create a json representation on your serialized model
+model_json = serializer.as_json
+```
+
+This will return an instance of the SerialzableResource class that will allow you to serialze your model.
```
Comment (beauby) [inline docs/howto/outside_controller_use.md]: This comment is a bit awkward here now that you added the final step :smile:

**Comment by CodedBeardedSignedTaylor** [comment]
True that. Deleted the redundant lines.

File: docs/README.md [hunk 6]
```diff
@@ -13,6 +13,8 @@ This is the documentation of AMS, it's focused on the **0.10.x version.**
 
 - [How to add root key](howto/add_root_key.md)
 
+- [Filtering Assocations](howto/filter_associations.md)
```
Comment (beauby) [inline docs/README.md]: Is this related to the current PR?

File: docs/howto/outside_controller_use.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+```ruby
+# Creating the resource
+post = Post.create(title: "Another Example", body: "So much fun.")
+
+# Optional options parameters
+options = {}
+
+# Create the serializer
+serializer = ActiveModel::Serializer.serializer_for(post, options)
```
Comment (bf4) [inline docs/howto/outside_controller_use.md]: Great job here!  you could also call `ActiveModel::SerializableResource.serialize(post, options).serializer` if you already have that, though it'll either give you ArraySerializer if the resource responds to :each, else the same as ` ActiveModel::Serializer.serializer_for(post, options)`

File: docs/README.md (hunk 6 above)
Comment (CodedBeardedSignedTaylor) [inline docs/README.md]: It is not! Good catch there!

File: docs/howto/outside_controller_use.md (hunk 7 above)
Comment (CodedBeardedSignedTaylor) [inline docs/howto/outside_controller_use.md]: Thanks! I'll be sure to include that option as well!

File: docs/howto/outside_controller_use.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+```ruby
+# Create our resource
+post = Post.create(title: "Sample post", body: "I love Active Model Serializers!")
+
+# Optional options parameters
+options = {}
+
+# Create the serializer
+serializer = ActiveModel::SerializableResource.serialize(post, options)
```
Comment (beauby) [inline docs/howto/outside_controller_use.md]: There's a terminology issue here. It is not specific to this PR, but `ActiveModel::SerializableResource.serialize()` actually returns an instance of `ActiveModel::SerializableResource` (it is equivalent to `ActiveModel::SerializableResource.new()`.
Hence, a possible replacement would be:
```ruby
serializable_resource = ActiveModel::SerializableResource.new(post, options)
model_json = serializable_resource.as_json
```

Comment (bf4) [inline docs/howto/outside_controller_use.md]: agreed :)

Comment (CodedBeardedSignedTaylor) [inline docs/howto/outside_controller_use.md]: Ahhhh. That definitely makes more sense. Probably going to go with that unless anyone wants to include both ways of doing it.

Comment (bf4) [inline docs/howto/outside_controller_use.md]: Yeah.. I've gone sour on `ActiveModel::SerializableResource.serialize` as a method with @beauby's urgings :) e.g. https://github.com/rails-api/active_model_serializers/issues/1098

**Comment by CodedBeardedSignedTaylor** [comment]
Updated a few things. Let me know if I missed something!

File: docs/howto/outside_controller_use.md [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+# Create the serializer
+serializer = ActiveModel::Serializer.serializer_for(post, options)
+```
+
+This will return an instance, if any, of the resource's serializer.
+
+You could also retrieve the serializer via: 
+
+```ruby
+ActiveModel::SerializableResource.serialize(post, options).serializer
```
Comment (bf4) [inline docs/howto/outside_controller_use.md]: Due to likely changes, I think we'll want `ActiveModel::SerializableResource.new(post, options).serializer`

Comment (bf4) [inline docs/howto/outside_controller_use.md]: cc @beauby

File: docs/howto/outside_controller_use.md (hunk 7 above)
Comment (bf4) [inline docs/howto/outside_controller_use.md]: Just to clarify, this is the serializer class i.e. `serializer == PostSerializer`, not the serializer instance (same below).

ref: https://github.com/rails-api/active_model_serializers/blob/64168cbecd6dfff99d81e4ab5444b36ffe219beb/lib/active_model/serializable_resource.rb#L46-L65

**Comment by bf4** [comment]
:+1:

**Comment by bf4** [comment]
(You might want to rebase...)

File: docs/howto/outside_controller_use.md (hunk 9 above)
Comment (beauby) [inline docs/howto/outside_controller_use.md]: This makes me very happy.

File: docs/howto/outside_controller_use.md (hunk 7 above)
Comment (beauby) [inline docs/howto/outside_controller_use.md]: Agreed, the comment "Create the serializer" might be misleading. What about "Retrieve the default serializer for `Posts`"?

**Comment by CodedBeardedSignedTaylor** [comment]
True that.

**Comment by CodedBeardedSignedTaylor** [comment]
Did somewhat of a clumsy rebase, but I believe it should be up to date now!

**Comment by CodedBeardedSignedTaylor** [comment]
Hmmmm. Build is failing for whatever reason. Looking into it now.

File: docs/howto/outside_controller_use.md (hunk 9 above)
Comment (joaomdmoura) [inline docs/howto/outside_controller_use.md]: @CodedBeardedSignedTaylor indeed if you are not passing a block you also don't need to use `.serialize` and can go straight with `ActiveModel::SerializableResource.new(post, options).serializer`

**Comment by joaomdmoura** [comment]
This is nice, something we're really looking forward to have, can you check the last comment @CodedBeardedSignedTaylor and squash the commits so I can merge it :smile: ?

File: docs/howto/outside_controller_use.md (hunk 9 above)
Comment (bf4) [inline docs/howto/outside_controller_use.md]: I think we're going to soon be removing the `.serialize` method, since it turns out it is more confusing than useful.

so, like @joaomdmoura wrote `serializable_resource = ActiveModel::SerializableResource.new(post, options).serializer `

(when I wrote the method, I thought the class would be re-used for deserialization..)

File: docs/howto/outside_controller_use.md [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+# Create a serializable resource instance
+serializable_resource = ActiveModel::SerializableResource.new(post, options)
+
+# Convert your resource into json
+model_json = serializable_resource.as_json
+```
+
+### Retrieving a Resource's Active Model Serializer
+
+If you want to retrieve a serializer for a specific resource, you can do the following:
```
Comment (bf4) [inline docs/howto/outside_controller_use.md]: I'm not sure what the use case is, right now, for getting a **serializer** class *or* instance.  The only thing I can think of is to test that the serializer works as expected, which still requires and adapter...

If you can think of one, I think it would be good to add

File: docs/howto/outside_controller_use.md [hunk 11]
```diff
@@ -0,0 +1,43 @@
+## Using AMS Outside Of A Controller
+
+### Serializing a resource 
+
+In the current version of AMS, serializing resources outside controllers is
+fairly simple:
```
Comment (bf4) [inline docs/howto/outside_controller_use.md]: maybe instead of `current version` -> `0.10 master`?

File: docs/howto/outside_controller_use.md [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+```ruby
+# Create our resource
+post = Post.create(title: "Sample post", body: "I love Active Model Serializers!")
+
+# Optional options parameters
+options = {}
```
Comment (bf4) [inline docs/howto/outside_controller_use.md]: TODO: link to description of `options` and how they are used ( as serializer_options, adapter_options, and renderer options)

File: docs/howto/outside_controller_use.md [hunk 13]
```diff
@@ -0,0 +1,43 @@
+## Using AMS Outside Of A Controller
+
+### Serializing a resource 
+
+In the current version of AMS, serializing resources outside controllers is
+fairly simple:
+
+```ruby
+# Create our resource
+post = Post.create(title: "Sample post", body: "I love Active Model Serializers!")
```
Comment (bf4) [inline docs/howto/outside_controller_use.md]: TODO (in future PRs):

- [ ] example resource doesn't need to be persisted or even an ActiveRecord object
  - [ ] unless your unpersisted resources needs an attribute only available in the db
  - [ ] unless your 'poro' needs complicated associations
- [ ] Show [ActiveModel::Serializer::Lint::Tests](https://github.com/rails-api/active_model_serializers/blob/1388ae82f26133d62f4e6cdfc40d1c659e121935/lib/active_model/serializer/lint.rb) and [usage](https://github.com/rails-api/active_model_serializers/blob/1388ae82f26133d62f4e6cdfc40d1c659e121935/test/poro_test.rb)

File: docs/howto/outside_controller_use.md (hunk 10 above)
Comment (bf4) [inline docs/howto/outside_controller_use.md]: Besides that, I'm :100: LGTM

**Comment by bf4** [comment]
@CodedBeardedSignedTaylor Lemme know if you'd like to pair on bringing this to completion, e.g. rebasing. I'm at http://benjaminfleischer.com/pair

File: docs/howto/outside_controller_use.md (hunk 10 above)
Comment (CodedBeardedSignedTaylor) [inline docs/howto/outside_controller_use.md]: I'm not really sure. I mainly added this because it was one of the requests of the initial issue. Still might be worth having for the moment.

File: docs/howto/outside_controller_use.md (hunk 12 above)
Comment (CodedBeardedSignedTaylor) [inline docs/howto/outside_controller_use.md]: Agreed.

File: docs/howto/outside_controller_use.md (hunk 13 above)
Comment (CodedBeardedSignedTaylor) [inline docs/howto/outside_controller_use.md]: (thumbsup)

**Comment by CodedBeardedSignedTaylor** [comment]
Pretty sure I squashed the commits this time. Need wait for the build to finish and we should be good!

**Comment by bf4** [comment]
TODOs in future PRs:

[Serializing a resource section](https://github.com/rails-api/active_model_serializers/pull/1083#discussion_r38998233) [2](https://github.com/rails-api/active_model_serializers/pull/1083#discussion_r38997895)

- [ ] example resource doesn't need to be persisted or even an ActiveRecord object
  - [ ] unless your unpersisted resources needs an attribute only available in the db
  - [ ] unless your 'poro' needs complicated associations
- [ ] Show [ActiveModel::Serializer::Lint::Tests](https://github.com/rails-api/active_model_serializers/blob/1388ae82f26133d62f4e6cdfc40d1c659e121935/lib/active_model/serializer/lint.rb) and [usage](https://github.com/rails-api/active_model_serializers/blob/1388ae82f26133d62f4e6cdfc40d1c659e121935/test/poro_test.rb)
- [ ] link to description of `options` and how they are used ( as serializer_options, adapter_options, and renderer options)

**Comment by CodedBeardedSignedTaylor** [comment]
Thanks for the offer @bf4! I think I did it correctly this time. Might need to work with you on future stuff though.

**Comment by CodedBeardedSignedTaylor** [comment]
Also, agreed to the To-DO stuff!

**Comment by bf4** [comment]
@CodedBeardedSignedTaylor are you in the AMS slack? https://amserializers.herokuapp.com/

**Comment by CodedBeardedSignedTaylor** [comment]
@bf4 Yep!

**Comment by joaomdmoura** [comment]
Can you squash it into a single commit please? :grin: @CodedBeardedSignedTaylor
Btw nice work here, pretty solid really want to merge this :smile:

**Comment by CodedBeardedSignedTaylor** [comment]
Did that work? I still have merges of the master's repo in my master (thus, they show up as commits). Sorry, I'm kind of bad at squashing commits :(

**Comment by NullVoxPopuli** [comment]
hey @CodedBeardedSignedTaylor if you could squash this PR, or re-submit a new PR with a single commit, that would be fantastic.

we'll give ya a couple days -- but after that, we'll do it for ya. But we'll be sure to set the commit author to you.

**Comment by CodedBeardedSignedTaylor** [comment]
#1155 is now the current pull request. Sorry for all the messiness!
